### ISSUE TITLE
False positive for no-member when using assignment expression (with torch?)

### ISSUE THREAD

**Issue by ringohoffman** [PyCQA/pylint#8178]
### Bug description

Consider the 2 functionally equivalent pieces of code:

```python
import torch
from torch import nn


def okay(module_or_parameter: nn.Parameter | torch.Tensor) -> None:
    if isinstance(module_or_parameter, nn.Parameter):
        param = module_or_parameter
        param.requires_grad = False
    elif isinstance(module_or_parameter, nn.Module):
        module = module_or_parameter
        module.eval()

        for param in module.parameters():
            param.requires_grad = False


def no_member_error(module_or_parameter: nn.Parameter | torch.Tensor) -> None:
    if isinstance((parameter := module_or_parameter), nn.Parameter):
        parameter.requires_grad = False
    elif isinstance((module := module_or_parameter), nn.Module):
        module.eval()  # pylint(no-member): Class 'module' has no 'eval' member

        for param in module.parameters():  # pylint(no-member): Class 'module' has no 'parameters' member
            param.requires_grad = False

```

### Command used

```shell
pylint a.py
```

### Pylint output

```shell
$ pylint a.py
************* Module a
a.py:21:8: E1101: Class 'module' has no 'eval' member (no-member)
a.py:23:21: E1101: Class 'module' has no 'parameters' member (no-member)
```

### Expected behavior

I expect not to get `no-member` errors.

### Pylint version

```shell
$ pylint --version
pylint 2.16.0
astroid 2.14.1
Python 3.10.6 (main, Oct 24 2022, 11:04:34) [Clang 12.0.0 ]
```

### OS / Environment

macOS 13.1 (22C65)

### Additional dependencies

torch==1.13.1
